File: App/Resources/AlgorithmDetails/pigeonholesort/pigeonholesort.cs
using System;
using System.Linq;

public class PigeonholeSort {
  public static int[] Sort(int[] array) {
    int min = array.Min();
    int max = array.Max();
    int size = max - min + 1;

    var holes = new int[size];
    foreach (var value in array) {
      holes[value - min]++;
    }

    var output = new int[array.Length];
    var j = 0;
    for (var count = 0; count < size; count++) {
      while (holes[count] > 0) {
        holes[count]--;
        output[j] = count + min;
        j++;
      }
    }
    return output;
  }

  public static void Main(String[] args) {
    int[] array = {0, 39, 21, 62, 91, 77, 14, 23,
      90, 69, 51, 81, 68, 83, 32, 56};
    array = Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}

File: App/Resources/AlgorithmDetails/pigeonholesort/pigeonholesort.js
function sort(arr) {
    var n = arr.length;
    var min = arr[0];
    var max = arr[0];
    for (var i = 1; i < n; i++) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }

    var size = max - min + 1;
    var holes = new Array(size).fill(0);
    for (var i = 0; i < n; i++) {
        holes[arr[i] - min]++;
    }

    var j = 0;
    for (var count = 0; count < size; count++) {
        while (holes[count] > 0) {
            holes[count]--;
            arr[j] = count + min;
            j++;
        }
    }
    return arr;
}

var array = [0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56];
sort(array);
console.log("[" + array.join(", ") + "]");
